File: frontend/src/lib/bankFeed.ts
import { apiGet, apiPost } from "@/lib/api";

export type BankFeedStatus = "MATCHED" | "UNMATCHED";

export type BankFeedTransaction = {
  id: number;
  bank_account_id: number;
  posted_date: string;
  amount: string;
  payee: string | null;
  description: string | null;
  memo: string | null;
  reference_number: string | null;
  source_provider: string;
  external_id: string | null;
  status: BankFeedStatus;
  matched_source_type: string | null;
  matched_source_id: number | null;
  created_at: string;
};

export type BankFeedList = {
  items: BankFeedTransaction[];
  total: number;
};

export type BankFeedImportResult = {
  imported: number;
  duplicates: number;
  matched: number;
  total: number;
};

export type BankFeedRematchResult = {
  matched: number;
  remaining_unmatched: number;
};

export function listBankFeed(
  bankAccountId: number,
  status?: BankFeedStatus
): Promise<BankFeedList> {
  const query = status ? `?status_filter=${encodeURIComponent(status)}` : "";
  return apiGet(
    `/api/accounting/bank-accounts/${bankAccountId}/bank-feed${query}`
  ) as Promise<BankFeedList>;
}

export function importBankFeed(
  bankAccountId: number,
  content: string
): Promise<BankFeedImportResult> {
  return apiPost(
    `/api/accounting/bank-accounts/${bankAccountId}/bank-feed/import`,
    { content }
  ) as Promise<BankFeedImportResult>;
}

export function rematchBankFeed(
  bankAccountId: number
): Promise<BankFeedRematchResult> {
  return apiPost(
    `/api/accounting/bank-accounts/${bankAccountId}/bank-feed/rematch`
  ) as Promise<BankFeedRematchResult>;
}

File: backend/app/schemas/bank_feed.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Literal, Optional

from pydantic import BaseModel, Field


class BankFeedImportIn(BaseModel):
    content: str = Field(..., min_length=1, max_length=5_000_000)


class BankFeedTransactionOut(BaseModel):
    id: int
    bank_account_id: int
    posted_date: date
    amount: Decimal
    payee: Optional[str] = None
    description: Optional[str] = None
    memo: Optional[str] = None
    reference_number: Optional[str] = None
    source_provider: str
    external_id: Optional[str] = None
    status: Literal["MATCHED", "UNMATCHED"]
    matched_source_type: Optional[str] = None
    matched_source_id: Optional[int] = None
    created_at: datetime


class BankFeedListOut(BaseModel):
    items: list[BankFeedTransactionOut]
    total: int


class BankFeedImportOut(BaseModel):
    imported: int
    duplicates: int
    matched: int
    total: int


class BankFeedRematchOut(BaseModel):
    matched: int
    remaining_unmatched: int
